metals/src/main/scala/scala/meta/internal/decorations/SyntheticDecorationProvider.scala

```diff
@@ -0,0 +1,240 @@
+package scala.meta.internal.decorations
+
+import java.nio.charset.Charset
+import java.nio.file.Path
+import java.util.concurrent.atomic.AtomicReference
+
+import scala.concurrent.ExecutionContext
+import scala.concurrent.Future
+
+import scala.meta.internal.metals.Buffers
+import scala.meta.internal.metals.ClientConfiguration
+import scala.meta.internal.metals.Diagnostics
+import scala.meta.internal.metals.MetalsEnrichments._
+import scala.meta.internal.metals.UserConfiguration
+import scala.meta.internal.mtags.Md5Fingerprints
+import scala.meta.internal.mtags.SemanticdbClasspath
+import scala.meta.internal.mtags.Semanticdbs
+import scala.meta.internal.semanticdb.Scala._
+import scala.meta.internal.semanticdb.TextDocument
+import scala.meta.internal.semanticdb.TextDocuments
+import scala.meta.io.AbsolutePath
+
+import org.eclipse.{lsp4j => l}
+
+final class SyntheticDecorationProvider(
+    workspace: AbsolutePath,
+    semanticdbs: Semanticdbs,
+    buffer: Buffers,
+    client: DecorationClient,
+    fingerprints: Md5Fingerprints,
+    charset: Charset,
+    diagnostics: Diagnostics,
+    focusedDocument: () => Option[AbsolutePath],
+    clientConfig: ClientConfiguration,
+    userConfig: () => UserConfiguration
+)(implicit ec: ExecutionContext) {
+
+  import SemanticdbTreePrinter.printSyntheticInfo
+
+  private object Document {
+    /* We update it with each compilation in order not read the same file on
+     * each change. When typing documents stay the same most of the time.
+     */
+    private val document = new AtomicReference[TextDocument]
+
+    def currentDocument: Option[TextDocument] = Option(document.get())
+
+    def set(doc: TextDocument): Unit = document.set(doc)
+  }
+
+  def publishSynthetics(path: AbsolutePath): Future[Unit] = {
+
+    // If focused document is not supported we can't be sure what is currently open
+    def isFocusedDocument =
+      focusedDocument().contains(path) || !clientConfig.isDidFocusProvider()
+
+    /**
+     * Worksheets currently do not use semanticdb, which is why this will not work.
+     * If at any time worksheets will support it, we need to make sure that the
+     * evaluation will not be replaced with implicit decorations.
+     */
+    if (!path.isWorksheet && isFocusedDocument) {
+      val textDocument = Document.currentDocument match {
+        case Some(doc) if workspace.resolve(doc.uri) == path => Some(doc)
+        case _ =>
+          val textDocument = semanticdbs
+            .textDocument(path)
+            .documentIncludingStale
+          textDocument.foreach(Document.set)
+          textDocument
+      }
+      textDocument match {
+        case Some(doc) =>
+          publishImplicitInfo(path, doc)
```

Comment:
```suggestion
          publishSyntheticInfo(path, doc)
```
This isn't just publishing implicits right, but other synthetics?

Reply:
Changed the name and actually dropped one method after removing diagnostics usage